src/lib.rs

```diff
@@ -112,31 +112,70 @@ pub use range::Range;
 fn err(e: ssize_t) -> io::Result<usize> {
     if e == -1 as ssize_t {
         Err(io::Error::last_os_error())
+    } else if e < 0 as ssize_t {
+        Err(io::Error::from(io::ErrorKind::InvalidData))
     } else {
         Ok(e as usize)
     }
 }
 
 fn read_at(fd: &RawFd, pos: u64, buf: &mut [u8]) -> io::Result<usize> {
-    err(unsafe {
-        pread(
-            *fd,
-            buf.as_mut_ptr() as *mut c_void,
-            buf.len() as size_t,
-            pos as off_t,
-        )
-    })
+    let mut total = 0usize;
+    let mut tries = 0usize;
+
+    while total < buf.len() && tries < 16 {
+        let bytes =
+            try!(
+                err(unsafe {
+                    pread(
+                        *fd,
+                        buf.as_mut_ptr().offset(total as isize) as *mut c_void,
+                        (buf.len() - total) as size_t,
+                        (pos + total as u64) as off_t,
+                )
+            }));
+
+        total += bytes;
```

Comment:
You must first check that `bytes >= 0`.  If `bytes == 0` that's EOF, so you should break out.  If it's `< 0` then you've encountered an error.  But it's really quite critical that you not add bytes to total before checking that it's non-negative.